PL2/PL2-parser/src/main/java/nl/tudelft/pl2016gr2/parser/controller/FullGfaReader.java

```diff
@@ -3,153 +3,218 @@
-   * @param graphsize the size of the graph.
+   * @param fileName the name of the file to read.
    */
-  public FullGfaReader(String filename, int graphsize) {
-    this.numNodes = graphsize;
-    this.sc = new Scanner(FullGfaReader.class.getClassLoader().getResourceAsStream(filename));
-    originalGraph = new OriginalGraph();
-    prepNodes(originalGraph, numNodes);
-    read();
-
-    sc.close();
-    sc = null;
+  public FullGfaReader(String fileName) {
+    read(fileName);
   }
 
   /**
-   * Initialize the nodes of a graph.
+   * Read a GFA file.
    *
-   * @param graph the graph of which to initialize a set amount of node.
-   * @param nodes the amount of nodes to initialize.
+   * @param fileName the name of the GFA file.
    */
-  protected static void prepNodes(OriginalGraph graph, int nodes) {
-    for (int i = 1; i < nodes + 1; i++) {
-      graph.addNode(new Node(i, 1, new ArrayList<>(), 0));
+  private void read(String fileName) {
+    try {
+      BufferedReader br = new BufferedReader(new InputStreamReader(FullGfaReader.class
+          .getClassLoader().getResourceAsStream(fileName)));
+      br.readLine();
+      String line;
+      while ((line = br.readLine()) != null) {
+        char[] chars = line.toCharArray();
+        switch (chars[0]) {
+          case 'L':
+            parseEdge(chars);
+            break;
+          case 'S':
+            parseNode(chars);
+            break;
+          case 'H':
+            parseHeader(chars);
+            break;
+          default:
+        }
+      }
+    } catch (IOException ex) {
+      Logger.getLogger(FullGfaReader.class.getName()).log(Level.SEVERE, null, ex);
     }
   }
-  
+
   /**
-   * Read the actual data.
+   * Parse a line of the GFA file which contains an edge.
+   *
+   * @param chars characters of the line.
    */
-  private void read() {
-    int count = 0;
-    while (sc.hasNext()) {
-      count++;
-      String next = sc.next();
-      switch (next) {
-        case "H":
-          handleH(sc);
-          break;
-        case "S":
-          handleS(sc);
-          break;
-        case "L":
-          handleL(sc);
-          break;
-        default:
-          System.out.println("Not S, H or L at: " + count);
-          sc.nextLine();
-      }
+  private void parseEdge(char[] chars) {
+    int from = 0;
+    int index = 2;
+    while (chars[index] != '\t') {
+      from *= SHIFT_BY_BASE_10;
+      from += chars[index++] - '0';
     }
-    System.out.println("Number of lines read: " + count);
+    index += 3;
+    int to = 0;
+    while (chars[index] != '\t') {
+      to *= SHIFT_BY_BASE_10;
+      to += chars[index++] - '0';
+    }
+    getNode(to).addInlink(from);
+    getNode(from).addOutlink(to);
   }
 
   /**
-   * Private method for handling an "H" as first character on the line of a gfa file.
-   * @param sc The scanner which is used to parse the file.
+   * Parse a line of the GFA file which contains a node.
```

Comment:
`which -> that`

' (which/that) contains a node.' is a restrictive clause. There are many lines in the GFA file, but this method only parses the ones that contain a node. It therefore restricts the definition of 'a line' in the main sentence.
Another way of looking at it: the clause is essential to the meaning of the sentence, leaving it out changes the definition (suddenly ALL, even non-node, lines will be parsed). If this is the case, you should use 'that'.

[which vs. that](http://www.getitwriteonline.com/archive/103103whichthat.htm)